drivers/gnss/gnss_u_blox_protocol/gnss_u_blox_protocol.c

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "gnss_u_blox_protocol.h"
+
+const uint32_t ubx_baudrate[UBX_BAUDRATE_COUNT] = {
+	4800,
+	9600,
+	19200,
+	38400,
+	57600,
+	115200,
+	230400,
+	460800,
+};
+
+static int ubx_validate_frame(uint16_t ubx_frame_size, uint8_t message_class, uint8_t message_id,
+			      uint16_t payload_size)
+{
+	if (ubx_frame_size > UBX_FRM_SZ_MAX ||
+	    ubx_frame_size < UBX_FRM_SZ_WITHOUT_PAYLOAD ||
+	    ubx_frame_size < UBX_FRM_SZ_WITHOUT_PAYLOAD + payload_size) {
+		return -1;
+	}
+
+	if (payload_size == 0) {
+		return 0;
+	}
+
+	uint16_t payload_size_expected;
+
+	switch (message_class) {
+	case UBX_CLASS_ACK:
+		switch (message_id) {
+		case UBX_ACK_ACK:
+			payload_size_expected = UBX_CFG_ACK_PAYLOAD_SZ;
+			break;
+		case UBX_ACK_NAK:
+			payload_size_expected = UBX_CFG_NAK_PAYLOAD_SZ;
+			break;
+		default: return -1;
+		}
+		break;
+	case UBX_CLASS_CFG:
+		switch (message_id) {
+		case UBX_CFG_RATE:
+			payload_size_expected = UBX_CFG_RATE_PAYLOAD_SZ;
+			break;
+		case UBX_CFG_PRT:
+			if (payload_size == UBX_CFG_PRT_POLL_PAYLOAD_SZ ||
+			    payload_size == UBX_CFG_PRT_SET_PAYLOAD_SZ) {
+				return 0;
+			} else {
+				return -1;
+			}
+			break;
+		case UBX_CFG_RST:
+			payload_size_expected = UBX_CFG_RST_PAYLOAD_SZ;
+			break;
+		case UBX_CFG_NAV5:
+			payload_size_expected = UBX_CFG_NAV5_PAYLOAD_SZ;
+			break;
+		case UBX_CFG_GNSS:
+			if ((payload_size - UBX_CFG_GNSS_PAYLOAD_INIT_SZ) %
+			    UBX_CFG_GNSS_PAYLOAD_CFG_BLK_SZ == 0) {
+				return 0;
+			} else {
+				return -1;
+			}
+			break;
+		case UBX_CFG_MSG:
+			payload_size_expected = UBX_CFG_MSG_PAYLOAD_SZ;
+			break;
+		default: return -1;
+		}
+		break;
+	default: return -1;
+	}
+
+	if (payload_size == payload_size_expected) {
+		return 0;
+	} else {
+		return -1;
+	}
+}
+
+int ubx_create_frame(uint8_t *ubx_frame, uint16_t ubx_frame_size, uint8_t message_class,
+		     uint8_t message_id, const void *data, uint16_t payload_size)
+{
+	if (ubx_validate_frame(ubx_frame_size, message_class, message_id, payload_size)) {
+		return -1;
+	}
+
+	struct ubx_frame_t *frame = (struct ubx_frame_t *) ubx_frame;
+
+	frame->preamble_sync_char_1 = UBX_PREAMBLE_SYNC_CHAR_1;
+	frame->preamble_sync_char_2 = UBX_PREAMBLE_SYNC_CHAR_2;
+	frame->message_class = message_class;
+	frame->message_id = message_id;
+	frame->payload_size_low = payload_size;
+	frame->payload_size_high = payload_size >> 8;
+
+	memcpy(frame->payload_and_checksum, (uint8_t *) data, payload_size);
+
+	uint16_t ubx_frame_len = payload_size + UBX_FRM_SZ_WITHOUT_PAYLOAD;
+
+	uint8_t ckA = 0, ckB = 0;
+
+	for (unsigned int i = UBX_FRM_CHECKSUM_START_IDX;
+	     i < (UBX_FRM_CHECKSUM_STOP_IDX(ubx_frame_len)); i++) {
+		ckA += ubx_frame[i];
+		ckB += ckA;
+	}
+
+	frame->payload_and_checksum[payload_size] = ckA;
+	frame->payload_and_checksum[payload_size + 1] = ckB;
+
+	return ubx_frame_len;
+}
```

Comment:
Hi @bjarki-trackunit, I have moved this to `modem_ubx.c`, but please note that payload size couldn't be validated in `modem_ubx.c` because details regarding gnss ublox messages is concern of `gnss_u_blox_protocol.c`. Hence, I have created a function in `ubx_create_and_validate_frame` in `gnss_u_blox_protocol.c`  that validates the payload size then calls `modem_ubx_create_frame` in `modem_ubx.c` to create the frame. Please check if this is correct.